pkg/k8s/endpoints.go

```diff
@@ -213,17 +235,33 @@ func parseEndpointPortV1Beta1(port slim_discovery_v1beta1.EndpointPort) (string,
 	return name, lbPort
 }
 
-// ParseEndpointSliceV1 parses a Kubernetes Endpoints resource
+// ParseEndpointSliceV1 parses a Kubernetes EndpointSlice resource.
+// It reads ready and terminating state of endpoints in the EndpointSlice to
+// return an EndpointSlice ID and a filtered list of Endpoints for service load-balancing.
 func ParseEndpointSliceV1(ep *slim_discovery_v1.EndpointSlice) (EndpointSliceID, *Endpoints) {
 	endpoints := newEndpoints()
 
 	for _, sub := range ep.Endpoints {
+		skipEndpoint := false
 		// ready indicates that this endpoint is prepared to receive traffic,
 		// according to whatever system is managing the endpoint. A nil value
 		// indicates an unknown state. In most cases consumers should interpret this
 		// unknown state as ready.
-		// More info: vendor/k8s.io/api/discovery/v1/types.go:117
+		// More info: vendor/k8s.io/api/discovery/v1/types.go
 		if sub.Conditions.Ready != nil && !*sub.Conditions.Ready {
+			skipEndpoint = true
+			if option.Config.EnableK8sTerminatingEndpoint {
+				// Terminating indicates that the endpoint is getting terminated. A
+				// nil values indicates an unknown state. Ready is never true when
+				// an endpoint is terminating. Propagate the terminating endpoint
```

Comment:
Kubernetes default behaviors allows graceful terminating, it never breaks established connections, doesn't matter the endpoints state.

This feature is meant to be used to allow sending NEW traffic to Terminating endpoints, "if there is no ready endpoint, then use the ones that are terminating AND SERVING"
```
	// serving is identical to ready except that it is set regardless of the
	// terminating state of endpoints. This condition should be set to true for
	// a ready endpoint that is terminating. If nil, consumers should defer to
	// the ready condition.
	// +optional
	Serving *bool
```

you should skip the endpoint is is terminating and not service

/cc @thockin